query_executor.go

```diff
@@ -83,7 +83,28 @@ func (q *queryExecutor) speculate(ctx context.Context, qry ExecutableQuery, sp S
 }
 
 func (q *queryExecutor) executeQuery(qry ExecutableQuery) (*Iter, error) {
-	hostIter := q.policy.Pick(qry)
+	type hostGetter interface {
```

Comment:
How about adding a private interface like:

```
type cqlQuery interface {
    ExecutableQuery
    getHost() *HostInfo
}
```

And make gocql internal code use `cqlQuery` instead of `ExecutableQuery` so that we don't have to worry about breaking the public `ExecutableQuery` interface when we add new functions to `Query` or `Batch`?

Reply:
Alternatively, we could just add new functions to `ExecutableQuery` because there's no place where a user can provide a custom implementation of `ExecutableQuery` to the driver is there? If the driver API doesn't expose any function that receives an object of type `ExecutableQuery` then we should be fine changing this interface since there's no real use case where a user might implement it on their application.

Reply:
From a quick glance at the code, `ExecutableQuery` is provided to implementations of `HostSelectionPolicy` but it's the driver that calls `HostSelectionPolicy.Pick(ExecutableQuery)` with `Query` or `Batch` objects as parameters so I don't see an issue in adding `GetHost` to `ExecutableQuery`

Reply:
I don't mind extending `ExecutableQuery` interface, but adding private methods is not the best thing to me. The driver exposes `HostSelectionPolicy` interface and allows user to provide their own implementations, but users cannot properly test them because `Pick()` accepts  `ExecutableQuery` that cannot be mocked for testing purposes due to private methods inside... Ofc they can just pass `Query` and `Batch` objects, but I'm not use if it is enough for testing

Reply:
These private methods would not be used in `HostSelectionPolicy` because `HostSelectionPolicy` declares the parameter as `ExecutableQuery` not `cqlQuery` or whatever the name would be for the private interface.

Reply:
A user would be able to continue testing custom implementations of `HostSelectionPolicy`

Reply:
> Alternatively, we could just add new functions to ExecutableQuery

I was thinking about the second approach

Reply:
But if we just add  `GetHost` to `ExecutableQuery` we don't need a private interface anyway so we can just go with that.

Reply:
But it still remains untestable because `ExecutableQuery` already has some private methods. But this is out of the scope of this PR. Ok, I'm going to update implementation with this one.

One thing came in mind, this PR doesn't provide this API for `Batch`. Should I add this API to `Batch` to be consistent with the `Query`? This PR focuses on virtual tables, and main purpose of Batch is data modification, so it actually cannot be used with virtual tables...

Reply:
I updated the implementation by adding `GetHost()` to `ExecutableQuery`.

Also, I looked over my implementation and found a bug leading to endless query execution... It appears when the host passed to `SetHost()` is not UP. I changed the behavior [here](https://github.com/worryg0d/gocql/blob/c1e46be251fc5779a2cb6a8a52cee03a8a161275/query_executor.go#L148) not to pick the next `selectedHost` when the selected one is not up and `GetHost()` returned no host.

Reply:
> But it still remains untestable because ExecutableQuery already has some private methods.

I know it's out of scope of this PR but I want to understand this, what private methods are you referencing?

Reply:
> One thing came in mind, this PR doesn't provide this API for Batch. Should I add this API to Batch to be consistent with the Query? This PR focuses on virtual tables, and main purpose of Batch is data modification, so it actually cannot be used with virtual tables...

I think it's fine if we add SetHost for query only, I don't see a use case for having it on batches atm

Reply:
>  what private methods are you referencing?

Such methods as `execute()`, `attempt()`, etc.

Reply:
Oh you're right, I didn't notice them. 👍 So it wouldn't actually make much of a difference if we added a private getHosts method there but we can keep it public